### ISSUE TITLE
Mocha can't run tests twice programmatically

### ISSUE THREAD

**Issue by aDu** [mochajs/mocha#2783]
Programmatically, I cannot run a Mocha test twice.

I.e. I can't do `mocha.run()` twice.

I looked at a similar issue (#[995](https://github.com/mochajs/mocha/issues/995)), however, the solutions are not optimal and/or not work.

For example, deleting the require cache is not a solution for me, as I need to use the require cache.

Is there a way to `mocha.run()` twice?

Thanks in advance!

**Comment by AlanDu50** [comment]
Mmmm ><. I guess I will stick with my solution of spawning a child process for each mocha run.

**Comment by KingRial** [comment]
Same problem here.
Another similar issue is #[736](https://github.com/mochajs/mocha/issues/736).
Both #995 and #736 didn't work for me ( I am still trying to find a different solution from spawning; I'll let you know if I find anything ).

**Comment by boneskull** [comment]
does the "retry" functionality not work?

**Comment by AlanDu50** [comment]
@boneskull I don't think so. I want to rerun the whole test suite, regardless if it was successful or not.

For example, maybe I want to run

**Comment by AlanDu50** [comment]
No, I don't think so. I will go with @KingRial 's solution

**Comment by josiahruddell** [comment]
I ran into the same problem. Why was this issue closed? Making a new instance of Mocha, and adding files should behave deterministically without modifying the require cache. Seems like a bug.

**Comment by AlanDu50** [comment]
@josiahruddell Good point, I agree.

**Comment by jcjolley** [comment]
+1

**Comment by ORESoftware** [comment]
+1

**Comment by rawb** [comment]
+1

**Comment by rawb** [comment]
`mocha.suite.suites = []`

**Comment by jamesmcging** [comment]
+1

**Comment by colinbendell** [comment]
Deleting the cache right after running require() seems to do nicely. I suspect it's very racy but I've been able to run the test hundreds of time in parallel.

```js
let mocha = new Mocha({});

mocha.suite.on('require', function (global, file) {
    delete require.cache[file];
});

//TODO: add files to suite & do things

mocha.run();
```

**Comment by colinbendell** [comment]
I've tested the selective delete of the require.cache with hundreds of parallel reruns. See my [comment on](https://github.com/mochajs/mocha/issues/995#issuecomment-365441585) #955. (this basically removes the test file from `require.cache` immediately after it is evaluated). The question is, why do you need a durable `require.cache`?

**Comment by nicojs** [comment]
For people interested, we've built the re-run functionality in the Stryker plugin for mocha: https://github.com/stryker-mutator/stryker/blob/930a7c39952e09a7d3f913fd9423bf78f608ec17/packages/mocha-runner/src/MochaTestRunner.ts#L48-L68

The problem we now face is that we _want to rerun mocha, without cleaning require cash or reloading the files_. We want to call it "hot reload". So rerun tests, without changing source code or test code.

Basically, what we want to do is this:

```js
const mocha = new Mocha(...);
mocha.addFile(...);
mocha.run(() => {
  mocha.run() => {
    console.log('ran twice');
  });
});
```

But right now, the second run always errors with: `"Cannot read property 'call' of undefined"` (which isn't helpful...)

The problem is that `mocha.run` has side effects. Namely: it cleans the test functions

here: https://github.com/mochajs/mocha/blob/0dacd1fb0067e40f8567653f828f677022e4fb89/lib/runner.js#L895-L897

and here: https://github.com/mochajs/mocha/blob/0dacd1fb0067e40f8567653f828f677022e4fb89/lib/suite.js#L455-L481

I think `mocha.run` isn't the place to start listening for the `suite end` event and clean up after. This should be done in the mocha cli file itself, right?

I'm willing to create the PR if the maintainers agree with this.

We could also add an option for this to keep the API backward compatible.

**Comment by plroebuck** [comment]
This is just _one_ aspect.. there are several others required to _really_ hope to reuse a Mocha instance.
Have something in mind, but post 6.0 release.

**Comment by plroebuck** [comment]
Of note, the Mocha-6.0 release API includes [`Mocha#unloadFiles`](https://github.com/mochajs/mocha/blob/52b5c42c3dda8c386735969642843bd1129a4562/lib/mocha.js#L356). It's one step towards eventual `mocha` reuse...

**Comment by nicojs** [comment]
Actually... no...

To delete the require cache correctly, not only the test files but also the _source files_ should be cleared. Mocha doesn't keep track of those files. This is what we need to do in Stryker right now (the `purgeFiles` method takes care of it).

Just to make sure we don't confuse 2 things: we actually want to reuse the same instance of mocha _without clearing any file cache_. Should I create a separate issue for that?

**Comment by boneskull** [comment]
essentially, you'd need to ensure Mocha's inner state was completely reset.  given there are bits of state everywhere, it's kind of a tall order at this point without some significant refactors.  granted, having a single source of truth for state would be a great refactor, but...nontrivial.

**Comment by nicojs** [comment]
> you'd need to ensure Mocha's inner state was completely reset. given there are bits of state everywhere, it's kind of a tall order at this point

Could you give me a list of things you can think about at the top of your head, with pointers on how to solve it? We should be able to create some kind of `state` object and pass that along. We can make it opt-in, so functionality remains the same if you don't want to use it.

**Comment by plroebuck** [comment]
> Could you give me a list of things you can think about at the top of your head, with pointers on how to solve it?

While ongoing efforts are made to payoff some of this technical debt for reusability, there are  higher priority issues to address (e.g., correctness, robustness). Would also need to redesign many of the automated tests to now run themselves twice. There can be no "opt-in" -- it would have to be baked-in.

**Comment by boneskull** [comment]
@nicojs Yes, what @plroebuck said, and that I can't explain Mocha's "state" strategy without saying "look at the code", because there's state everywhere.

**Comment by boneskull** [comment]
That said, I'm happy to entertain ideas/PRs to solve the "state" question, I just don't think our maintainer team has the resources to embark on such a ...quest.

**Comment by nicojs** [comment]
> I just don't think our maintainer team has the resources to embark on such a ...quest.

Haha 

Thanks for the clear responses. If I find a couple of hours in the coming weeks, I might take a look. One might say, I will embark on a quest 🏇

**Comment by DiogoMCampos** [comment]
> > I just don't think our maintainer team has the resources to embark on such a ...quest.
> 
> Haha
> 
> Thanks for the clear responses. If I find a couple of hours in the coming weeks, I might take a look. One might say, I will embark on a quest horse_racing

I don't feel comfortable on undertaking a nontrivial _quest_ without knowing the codebase, but making myself available to help if needed. This would solve a current major issue I have.

**Comment by boneskull** [comment]
I hate to pull the redux card, but something like redux might be a good solution here.

**Comment by plroebuck** [comment]
> I don't feel comfortable on undertaking a nontrivial _quest_ without knowing the codebase, but making myself available to help if needed.

[This](https://www.youtube.com/watch?v=--hMJPUBwMc) seems appropriate somehow.

**Comment by wKich** [comment]
Hello everyone. I'll try to write tool on top of mocha, to run heavy e2e tests in separate workers for each browser for example. Current solutions (like [mocha-parallel-tests](https://github.com/mocha-parallel/mocha-parallel-tests)) is not prefect for me, it don't allow share selenium webdriver connection between tests. So I want to start pool of workers, that each other initialize webdriver and all test cases, then master thread send message to worker for start specific test case and worker send result back on finish.
[`Mocha#unloadFiles`](
https://github.com/mochajs/mocha/blob/52b5c42c3dda8c386735969642843bd1129a4562/lib/mocha.js#L356), seems could help, but unload cache and reinitialize tests for each run it's little bit overwhelmed.

**Comment by wKich** [comment]
@boneskull, can you clarify, why `cleanReferences` remove only `fn` property from each hook instead of clear entire array?

**Comment by nicojs** [comment]
Ok, I've investigated the issues a bit and I want to propose the following changes. They are designed to be 100% backward compatible and minor while still allowing for the new feature. @boneskull @plroebuck do you agree with these changes? I'd be willing to prepare the PR.

## 🔀 `cleanReferences`

I want to add a feature to allow references not to be cleared. Let's call it `autoDispose`. By default it would be `true` (non-breaking change). But if you use mocha programmatically you are allowed to set it to `false`, you would have to call `dispose` on the mocha instance later if you still want to dispose.

## 🔢 "bits of state"

For the _bits of state_ issue, I'm pretty sure that the state is situated in `Suite`, `Test` and `Runnable`. I would like to add a `reset` method in each, which resets the state. It would be called from the `Mocha` class, whenever an _nth_ test run is started.

## 🏡 Housekeeping

I would also want to add 2 new validations:

1. Whenever mocha's `run` method is called an "object is already disposed" error will be thrown whenever the mocha's instance is already disposed (either by autoDispose, or by a manual dispose).
1. Whenever mocha's `run` method is called while a run is already in progress a "mocha run is already in progress" error is thrown.

**Comment by boneskull** [comment]
sounds reasonable to me. you’ll also want to ensure there’s no eventemitter leaks happening.

**Comment by nicojs** [comment]
The PR is finished. @boneskull would you care to take a look?

Great integration tests you guys have! I broke mocha's `watch` functionality apparently. Never would have known without those awesome tests!

I choose in the end NOT to call it `autoDispose`, rather i've called it `cleanReferencesAfterRun`. My reasoning is that `dispose()` on the mocha instance should also remove any `uncaughtException` listeners remaining on the `process`, but I don't want to remove the listeners directly after a test run, since the previous runner should still handle them (it was one of the integration tests). That's why I instead only clean all references after each run. You can still dispose everything at once using the `dispose` method on a mocha instance.

**Comment by RossVertizan** [comment]
Thank you @nicojs, this works perfectly for what I am trying to achieve (something similar to your need).

Quick question though, where do I set `cleanReferencesAfterRun`, at the moment, I have hacked it into runner.run(), but can I pass it as an option?  I see it is set on the Mocha object, but can't see how I would pass it to Mocha.

Many thanks for any insights.

**Comment by nicojs** [comment]
You can call it on the mocha object itself, using the programmatic interface: https://mochajs.org/api/mocha

@RossVertizan are you building a mutation testing framework by any chance? 😉

**Comment by RossVertizan** [comment]
Ahh OK, so I need to call the cleanReferencesAfterRun method on the Mocha object.  It's a shame it can't be passed as a Mocha option.

@boneskull @craigtaub  -  is there possibility we could add this as a Mocha option?

@nicojs - actually no not building a mutation testing framework.  I am building a constrained random testing engine for web and mobile app testing we call [Vitaq](https://vitaq.io/). We have a concept of action/page and next allowable action/page, so that the user can define a graph/model of their website/app and the tool will walk through the graph executing the script (suite in Mocha parlance) associated with each page.  We can define sequences or user paths through the model and set functional coverage targets for the sequences, each page individually and for the use of constrained random variables and we use AI to drive the functional coverage target to closure.

We can (and frequently do) have loops e.g. navigating back to the home page before going off somewhere else and we therefore need the ability to execute the suite multiple times and hence the need for your enhancement.

I am currently looking at making the tool work with the webdriverio infrastructure, using Mocha as the test framework.  So I don't have direct access to the Mocha object (it's in the webdriverio-mocha interface code) and It would be nice to be able to pass cleanReferencesAfterRun as a mocha option.  The passing of options to Mocha is already supported by webdriverio.

**Comment by gilad-moskowitz** [comment]
Can somebody please provide the syntax needed to programmatically run tests twice? Currently, I am having an issue where I am creating a Mocha instance, setting the cleanReferencesAfterRun : false, and when I run the tests again I get a 'done() called multiple times' error as the first run through had tests that failed due to timeout. I am not calling 'done()' anywhere since I am writing the tests using async/await so I don't know how to resolve the issue. The only thing I can think of is disposing of the mocha instance that ran the first suite of tests and creating a new one for the second suite, but if I call mocha.unloadFiles() or mocha.dispose(), my process ends. Any advice would be greatly appreciated.

**Comment by RossVertizan** [comment]
@gilad-moskowitz how are you "setting the cleanReferencesAfterRun : false" on the Mocha instance?
There is a method implemented to do this (called 'cleanReferencesAfterRun'), so you should call that method, is that what you are doing?

**Comment by nicojs** [comment]
For anyone interested in this API, here is a small example:

```js
const Mocha = require('mocha');

async function main(){
    const mocha = new Mocha();
    mocha.addFile('test/test.spec.js');
    mocha.cleanReferencesAfterRun(false); // => important! This allows multiple runs.
    await mocha.loadFilesAsync();
    mocha.run(() => {
        console.log('✅ done');
        mocha.run(() => {
            console.log('\t✅ done x2');
            mocha.dispose(); // => important if you want to keep the node process alive, otherwise memory leak 🕳
        });
    });
}

main().catch(err => {
    console.error(err);
    process.exitCode = 1;
});
```

This also works with async specs, see my test.spec.js:

```js

function sleep(n) {
    return new Promise(res => setTimeout(res, n));
}

it('should work', () => {})

it('should work async as well', () => sleep(100) );
```

**Comment by RossVertizan** [comment]
@nicojs Thanks - that's what I should have done.

**Comment by gilad-moskowitz** [comment]
Hello, thank you so much for the quick response. So, my code that returns the `done() called multiple times` error is as follows:
```
const Mocha = require('mocha');
const fs = require('fs')
const path = require('path')

async function main(){
    const mocha = new Mocha({
         ui: 'bdd',
         reporter: 'custom.js',
    });
    fs.readdirSync('test').filter(function(file) {
         return file.substr(-8) === '.test.js';
    }).forEach(function(file){
         mocha.addFile(path.join('test', file));
    });
    }
    await mocha.loadFilesAsync();
    mocha.cleanReferencesAfterRun(false).run(() => {
        baseTimeout = 2*baseTimeout;
        mocha.cleanReferencesAfterRun(true).run(() => {mocha.dispose();});
    });
}

main().catch(err => {
    console.error(err);
    process.exitCode = 1;
});
```

I set the timeout for each test as being equal to the baseTimeout before each test. So on the first run the tests have a timeout equal to the baseTimeout and on the second run the tests have a timeout equal to 2*baseTimeout. I use Chai for making API requests and all my tests are written as (for example):
```
describe('Suite name', function (){
    it('Test name', async function(){
         let result = await chai.request(API).get('...');
         await expect(result.status).to.equal(200);
    });
})
```

On the first run, sometimes the tests fail because the timeout. This causes tests on the second run to error with `done() called multiple times` even though all my code uses ASYNC/AWAIT so there is no `done()` call at all. 

My guess as to why this issue occurs is because the chai request that I am awaiting doesn't resolve until my second run and when it resolves it causes a problem.

**Comment by gilad-moskowitz** [comment]
@nicojs @RossVertizan Just wanted to follow up and see if you had a chance to look at my most recent comment. I haven't found any work arounds to the issue so any advice would be greatly appreciated.

**Comment by nicojs** [comment]
Feel free to upload a small reproduction repo, so I can try to reproduce this locally

**Comment by gilad-moskowitz** [comment]
@nicojs I created a private reproduction repository. You should be able to recreate it locally using the files there. Thank you so much for the help.

**Comment by gilad-moskowitz** [comment]
@nicojs Just wanted to follow up again and check to see if you had gotten the access to my reproduction repo. Thank you

**Comment by gabssnake** [comment]
When creating multiple instances of `Mocha` it *will* leak EventListeners attached to `process` `uncaughtException`.

This happens even if you don't manually set the `cleanReferencesAfterRun` option.

Calling `mocha.dispose()` solves the issue for me.
